contributor-advice.md

```diff
@@ -0,0 +1,107 @@
+
+This must be ensured by manually verifying the license for all the dependencies for the project:
+
+1. Whenever a committer to the project changes a version of a dependency (including Scala) in the build file.
+2. Whenever a committer to the project adds a new dependency.
+3. Whenever a new release is cut (public or private for a customer).
+
+Every external dependency listed in the build file must have a trailing comment with the license name of the dependency.
+
+Which licenses are compatible with Apache 2 are defined in [this doc](http://www.apache.org/legal/3party.html#category-a), where you can see that the licenses that are listed under ``Category A`` automatically compatible with Apache 2, while the ones listed under ``Category B`` needs additional action:
+
+> Each license in this category requires some degree of [reciprocity](http://www.apache.org/legal/3party.html#define-reciprocal); therefore, additional action must be taken in order to minimize the chance that a user of an Apache product will create a derivative work of a reciprocally-licensed portion of an Apache product without being aware of the applicable requirements.
+
+
+### Packages & Scoping
+
+Use `private`, `private[connector]` and `final` extensively to limit the API surface.
+
+| Package                                  | Purpose
+| -----------------------------------------|------------------------
+| `akka.stream.alpakka.connector.javadsl`  | Java-only part of the API, normally factories for Sources, Flows and Sinks
+| `akka.stream.alpakka.connector.scaladsl` | Scala-only part of the API, normally factories for Sources, Flows and Sinks
+| `akka.stream.alpakka.connector`          | Shared API, eg. settings classes, implementation if thorowly made private
+| `akka.stream.alpakka.connector.impl`     | Optional, implementation in separate package
+
+### Graph stage checklist
+
+* Keep mutable state within the `GraphStageLogic` only 
+* Open connections in `preStart`
+* Release resources in `postStop`
+* Fail early on configuration errors
+* No Blocking At Any Time 
+* Make sure the code is thread-safe
+
+### Keep the code DRY
+
+Avoid duplication of code between different Sources, Sinks and Flows. Extract the common logic to a common abstract
+base `GraphStageLogic` that is inherited by the `GraphStage`s.
+
+In some cases it may be useful to implement the Sources or the Sinks by means of Flows:
+* Source: `Source.maybe.viaMat(MyFlow)(Keep.right)`
+* Sink: `MyFlow.toMat(Sink.ignore)(Keep.right)`
+
+You do not need to expose every configuration a Flow offers this way -- Focus on the most important ones.
+
+
+## Test
+
+Write tests for all public methods and possible settings.
+
+Use Docker containers for any testing needed by your application. The container's configuration can be added to the
```

Comment:
In my view, before suggesting to use docker, we should encourage to use internal testing API like MiniSolrCloudCluster for Solr, ElasticsearchClusterRunner for Elasticsearch, if it is possible.

Reply:
Using Docker make the testing environment more isolated than if we configure stuff just for the test. It allows us to limit, memory, disk or whatever we want to limit. And it's trivial to update regardless of the connector version or dependencies.

Also, I'd rather pulling a container for testing than adding an extra dependency.